elements/rh-secondary-nav/rh-secondary-nav.ts

```diff
@@ -0,0 +1,389 @@
+   * @param element:
+   * @returns {boolean}
+   */
+  static isDropdown(element: Element|null): element is RhSecondaryNavDropdown {
+    return element instanceof RhSecondaryNavDropdown;
+  }
+
+  /**
+   * Updates isMobile property with matchMedia when viewport changes
+   */
+  protected matchMedia = new MatchMediaController(this, `(min-width: ${tabletLandscapeBreakpoint})`, {
+    onChange: ({ matches }) => {
+      this._isMobile = !matches;
+    }
+  });
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.addEventListener('change', this._dropdownChangeHandler as EventListener);
+    this.addEventListener('overlay-change', this._toggleNavOverlay as EventListener);
+    this.addEventListener('focusout', this._focusOutHandler);
+    this.removeAttribute('role');
+  }
+
+  firstUpdated() {
+    // after update the overlay should be available to attach an event listener to
+    this._overlay?.addEventListener('click', this._overlayClickHandler as EventListener);
+    this._isMobile = (window.outerWidth < parseInt(tabletLandscapeBreakpoint.toString().split('px')[0]));
+  }
+
+  render() {
+    const classes = { 'is-mobile': this._isMobile };
+
+    return html`
+    <nav part="base" class="${classMap(classes)}">
+      <slot name="base">
+        <div id="container" part="container">
+          <slot name="logo"></slot>
+          <button aria-controls="${this.id}" @click="${this.#toggleMobileMenu}">Menu</button>
+          <slot name="nav"></slot>
+          <div id="cta" part="cta">
+            <slot name="cta"><slot>
+          </div>
+        </div>
+      </slot>
+    </nav>
+    <rh-secondary-nav-overlay hidden></rh-secondary-nav-overlay>
+    `;
+  }
+
+  /**
+   * Public API, opens a specific dropdown based on index.
+   * Closes all open dropdowns before opening specified.
+   * Toggles overlay to open
+   * @param index
+   * @returns {void}
+   */
+  public open(index: number): void {
+    if (index == null) {
+      return;
+    }
+    const dropdown = this.#dropdownByIndex(index);
+    if (dropdown && RhSecondaryNav.isDropdown(dropdown)) {
+      this.close();
+      this.#expand(index);
+      dropdown?.querySelector('a')?.focus();
+      this._overlay?.toggleNavOverlay(this._overlay, true, this);
+    }
+  }
+
+  /**
+   * Public API, closes all open dropdowns
+   * @returns {void}
+   */
+  public close(): void {
+    const dropdowns = this.#allDropdowns();
+    dropdowns.forEach(dropdown => this.#closeDropdown(dropdown));
+  }
+
+  /**
+   * When dropdown event is triggered gets dropdown index that triggered
+   * event then closes all dropdowns.
+   * If the event is to open a dropdown, run #expand(index)
+   * If isMobile is set dispatch an SecondaryNavOverlayEvent event
+   * to open the overlay
+   * @param event - {SecondaryNavDropdownChangeEvent}
+   * @return {void}
+   */
+  @bound
+  private _dropdownChangeHandler(event: SecondaryNavDropdownChangeEvent): void {
```

Comment:
🤔 thinking this over

if you want to keep the parent in charge of setting state, would it make sense to call the event `expand-request`?

Reply:
Assuming this is also just a naming issue, I have no problem switching that to expand-request.  It does give the event a more specific meaning.  

As far as wanting the parent in charge,  I assumed the parent would be the best place to handle the state of its children's dropdowns.  More specifically that any new dropdown open event closes all other open dropdowns so I just followed the parent -> child model thought process. 

You had mentioned at some point the idea of possibly abstracting the dropdowns (flyout menus)  into a completely separate component.  I haven’t thought about how that change would impact how I'm currently handling the open-close here.  

I can potentially see dropdowns listening for other dropdowns open request events, and if one is open all others close in response.  For reusability, If they aren’t used in a “grouping” like in the navigation menu itself, maybe you might not want that functionality. You might want to allow more than 1 to be open at a time, such as in an accordion, maybe?  Thought for later.  

That said, would something like that approach be preferable?  The dropdowns listen for `expand-request` events, and if it wasn’t the dropdown that called the event it would change its own state to closed?

Reply:
Fixed in a14bf72a07b3519ddbe0c9e65ea0840f047dc53f